python/singa/optimizer.py

```diff
@@ -251,47 +287,70 @@ class AdaGrad(Optimizer):
-        grad = self.apply_regularizer_constraint(epoch, value, grad, name)
+    def apply_with_lr(self, epoch, lr, grad, value, name, rc=True):
+        if rc:
+            grad = self.apply_regularizer_constraint(epoch, value, grad, name)
         if name is not None and name in self.learning_rate_multiplier:
             lr = lr * self.learning_rate_multiplier[name]
         self.opt.Apply(epoch, lr,  name, grad.singa_tensor, value.singa_tensor)
         return value
 
 
-class RMSProp(Optimizer):
-    '''RMSProp optimizer.
+class Adam(Optimizer):
+    '''Adam optimizer.
 
     See the base Optimizer for all constructor args.
 
     Args:
-        rho (float): float within [0, 1]
+        beta_1(float): coefficient of momentum
+        beta_2(float): coefficient of aggregated squared gradient
         epsilon (float): small value for preventing numeric error
     '''
 
-    def __init__(self, rho=0.9, epsilon=1e-8, lr=None, weight_decay=None,
-                 lr_gen=None, regularizer=None, constraint=None):
-        super(RMSProp, self).__init__(lr, weight_decay, lr_gen, regularizer,
-                                      constraint)
-        conf = model_pb2.OptimizerConf()
-        conf.rho = rho
-        conf.delta = epsilon
-        self.opt = singa.CreateOptimizer('RMSProp')
-        self.opt.Setup(conf.SerializeToString())
+    def __init__(self, beta_1=0.9, beta_2=0.999, epsilon=1e-8, lr=None,
+                 weight_decay=None, lr_gen=None, regularizer=None,
+                 constraint=None):
+        super(Adam, self).__init__(lr, None, weight_decay, lr_gen, regularizer,
+                                   constraint)
+        self.beta_1 = beta_1
+        self.beta_2 = beta_2
+        self.epsilon = epsilon
+        self.m = {}
+        self.v = {}
 
-    def apply_with_lr(self, epoch, lr, grad, value, name):
-        grad = self.apply_regularizer_constraint(epoch, value, grad, name)
+    def apply_with_lr(self, step, lr, grad, value, name, rc=True):
+        '''Update one parameter object.
+
+        Args:
+            step(int): the accumulated training iterations (not the epoch)
+        '''
+        t = step + 1
+        if rc:
+            grad = self.apply_regularizer_constraint(step, value, grad, name)
         if name is not None and name in self.learning_rate_multiplier:
             lr = lr * self.learning_rate_multiplier[name]
-        self.opt.Apply(epoch, lr,  name, grad.singa_tensor, value.singa_tensor)
+        if name not in self.m or name not in self.v:
+            self.m[name] = tensor.Tensor(grad.shape, grad.device, grad.dtype)
+            self.m[name].set_value(0)
+            self.v[name] = tensor.Tensor(grad.shape, grad.device, grad.dtype)
+            self.v[name].set_value(0)
+
+        self.m[name] *= self.beta_1
+        tensor.axpy(1 - self.beta_1, grad, self.m[name])
+        self.v[name] *= self.beta_2
+        tensor.axpy(1 - self.beta_1, tensor.square(grad), self.v[name])
```

Comment:
typo: should be "tensor.axpy(1 - self.beta_2, tensor.square(grad), self.v[name])"